kubernetes/deployments_kubeclient.go

```diff
@@ -0,0 +1,1339 @@
+				duration := now.Sub(startTime)
+				if duration > containerTimeout {
+					return true, false
+				}
+			}
+		}
+	}
+
+	return false, false
+}
+
+func isPullingImage(pod *v1.Pod) bool {
+	const containerCreating string = "ContainerCreating"
+	// If pod is pending with a container waiting due to a "ContainerCreating" event,
+	// categorize as "Pulling". This may change as more information is made available.
+	// See: https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js#L663
+	if pod.Status.Phase == v1.PodPending {
+		for _, status := range pod.Status.ContainerStatuses {
+			waiting := status.State.Waiting
+			if waiting != nil && waiting.Reason == containerCreating {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isPodReady(pod *v1.Pod) bool {
+	// If all of the pod's containers have a ready status, then the pod is
+	// considered ready.
+	total := len(pod.Spec.Containers)
+	numReady := 0
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Ready {
+			numReady++
+		}
+	}
+	return numReady == total
+}
+
+func (kc *kubeClient) getBestRoute(namespace string, dc *deployment) (*url.URL, error) {
+	serviceMap, err := kc.getMatchingServices(namespace, dc)
+	if err != nil {
+		return nil, err
+	}
+	// Get routes and associate to services using spec.to.name
+	err = kc.getRoutesByService(namespace, serviceMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find route with highest score according to heuristics from web-console
+	var bestRoute *route
+	bestScore := -1
+	for _, routes := range serviceMap {
+		for _, route := range routes {
+			score := scoreRoute(route)
+			if score > bestScore {
+				bestScore = score
+				bestRoute = route
+			}
+		}
+	}
+
+	// Construct URL from best route
+	var result *url.URL
+	if bestRoute != nil {
+		scheme := "http"
+		if bestRoute.tls {
+			scheme = "https"
+		}
+		result = &url.URL{
+			Scheme: scheme,
+			Host:   bestRoute.host,
+			Path:   bestRoute.path,
+		}
+	}
+
+	return result, nil
+}
+
+func (kc *kubeClient) getMatchingServices(namespace string, dc *deployment) (routesByService map[string][]*route, err error) {
+	services, err := kc.Services(namespace).List(metaV1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Check if each service's selector matches labels in deployment's pod template
+	template := dc.current.Spec.Template
+	if template == nil {
+		return nil, errors.New("No pod template for current deployment")
+	}
+	routesByService = make(map[string][]*route)
+	for _, service := range services.Items {
+		selector := service.Spec.Selector
+		match := true
+		// Treat empty selector as not matching
+		if len(selector) == 0 {
+			match = false
+		}
+		for key := range selector {
+			if selector[key] != template.Labels[key] {
+				match = false
+			}
+		}
+		// If all selector labels match those in the pod template, add service key to map
+		if match {
+			routesByService[service.Name] = nil
```

Comment:
why do you use a map for the keys only (ie, not fill any value for the service name key) ?

Reply:
The map's keys are filled by this method, which we use to keep track of the services that we want routes for. The map's values are filled in by the getRoutesByService method called just after this one. I'll change the initialization quoted above to be an empty slice instead of nil and add an additional comment, which should be more readable.

Reply:
ok